Refact documentation to leverage jupytext

SMoraisAnsys opened the pull request:
Another WIP in parallel of #3940 to test a new approach and keep the old one untouched ftm.

jorgepiloto commented on doc/source/conf.py:
```diff
@@ -77,9 +80,27 @@ def remove_doctree(app, exception):
+    """Copy directory examples (root directory) files into the doc/source/examples directory.
+    """
+    DESTINATION_DIRECTORY = pathlib.Path(app.srcdir, "examples").resolve()
+    print(f"Copying examples from {EXAMPLES_DIRECTORY} to {DESTINATION_DIRECTORY} directory")
```
It would be great to use here the logging from Sphinx. This way, this message would be collected in the log files and during the output.

jorgepiloto replied:
https://www.sphinx-doc.org/en/master/extdev/logging.html

SMoraisAnsys commented:
@jorgepiloto @dcrawforAtAnsys  In the end I played a bit with the hooks and I added two functions to:
- add ipython magic %%time to every code cell in the sources used by sphinx (this leave our examples clean while allowing us to retrieve the execution time of each cell)
- remove ipython magic %%time from the html files which could be confusing for our users.

This is the resulting html page
![image](https://github.com/ansys/pyaedt/assets/146729917/94926dfb-ee7a-4ae6-bf0f-9467349911d5)

dcrawforAtAnsys commented:
@SMoraisAnsys , @jorgepiloto : This looks good, but I cannot build local documentation. There seem to be import errors for PyAedt. Further investigation is needed.

dcrawforAtAnsys commented:
> @SMoraisAnsys , @jorgepiloto : This looks good, but I cannot build local documentation. There seem to be import errors for PyAedt. Further investigation is needed.

Cleaning the build folder fixed the problem :)

dcrawforAtAnsys commented:
@SMoraisAnsys : There are two issues that I see:
1. The relative path to the image doesn't work. This could be solved by putting _static content into the Examples folder.
2. Equations are not rendered at all. See the updated ``00_EDB_Create_VIA.py``. Still need to investigate this but I think we're on the right path.

PipKat commented on doc/source/conf.py:
```diff
@@ -173,6 +173,26 @@ def add_ipython_time(app, docname, source):
     # logger.info(source[0])
 
+def adjust_image_path(app, docname, source):
+    """Adjust the HTML label used to insert images in our examples.
```
```suggestion
    """Adjust the HTML label used to insert images in the examples.
```
Avoid "our" and "we".

PipKat commented on doc/source/conf.py:
```diff
@@ -173,6 +173,26 @@ def add_ipython_time(app, docname, source):
+def adjust_image_path(app, docname, source):
+    """Adjust the HTML label used to insert images in our examples.
+    
+    The following path makes the examples in the root directory working
```
```suggestion
    The following path makes the examples in the root directory work:
```

PipKat commented on doc/source/conf.py:
```diff
@@ -173,6 +173,26 @@ def add_ipython_time(app, docname, source):
+    
+    The following path makes the examples in the root directory working
+    # <img src="../../doc/source/_static/diff_via.png" width="500">
+    However, it fails when used through the documentation build since
```
```suggestion
    However, examples fail when used through the documentation build because
```

PipKat commented on doc/source/conf.py:
```diff
@@ -173,6 +173,26 @@ def add_ipython_time(app, docname, source):
+    # <img src="../../doc/source/_static/diff_via.png" width="500">
+    However, it fails when used through the documentation build since
+    reaching the associated path should be "../../_static/diff_via.png".
+    Indeed, directory _static is automatically copied in output directory
```
```suggestion
    Indeed, the directory ``_static`` is automatically copied into the output directory
```

PipKat commented on doc/source/conf.py:
```diff
@@ -173,6 +173,26 @@ def add_ipython_time(app, docname, source):
+    However, it fails when used through the documentation build since
+    reaching the associated path should be "../../_static/diff_via.png".
+    Indeed, directory _static is automatically copied in output directory
+    _build/html/_static.
```
```suggestion
    ``_build/html/_static``.
```

PipKat commented on doc/source/conf.py:
```diff
@@ -173,6 +173,26 @@ def add_ipython_time(app, docname, source):
+    # Get the full path to the document
+    docpath = os.path.join(app.srcdir, docname)
+
+    # Check if this is a .py example file
```
```suggestion
    # Check if this is a PY example file
```

PipKat commented on doc/source/conf.py:
```diff
@@ -173,6 +173,26 @@ def add_ipython_time(app, docname, source):
+    if not os.path.exists(docpath + '.py') or not docname.startswith("examples"):
+        return
+
+    logger.info(f"Changing HTML image path in file {docname}.py")
```
```suggestion
    logger.info(f"Changing HTML image path in the '{docname}.py' file.")
```
Not sure the single quotes are supported, but if they are....

dcrawforAtAnsys commented:
@SMoraisAnsys The image copying works. *.py files are still not rendered in HTML as they are in Jupyter Lab using Jupytext. I added some examples in the folder ``00-EDB`` and also created a ``_static`` folder.

jorgepiloto commented:
I would avoid centering the output of the cells. This breaks the custom formatting that an input cell may impose.

If it is fine by you, I will upload in a single commit the friendly CSS style that removes the cell numbers and ensures the proper space and font size.

jorgepiloto commented:
> I would avoid centering the output of the cells. This breaks the custom formatting that an input cell may impose.
> 
> If it is fine by you, I will upload in a single commit the friendly CSS style that removes the cell numbers and ensures the proper space and font size.

I just collected all CSS files under `doc/source/_static/css` in https://github.com/ansys/pyaedt/pull/4064/commits/1ba25b6807e5bf27dc1c954f7740283bf10e7752

SMoraisAnsys commented:
@dcrawforAtAnsys I changed the modification you did to a previous commit. 

I don't have issues with equations, see

![image](https://github.com/ansys/pyaedt/assets/146729917/3edbaeb6-4473-4c3b-9b67-4e56b679f641)

and

![image](https://github.com/ansys/pyaedt/assets/146729917/d661eb82-e851-4f56-9762-82b3d1589044)

> @SMoraisAnsys *.py files are still not rendered in HTML as they are in Jupyter Lab using Jupytext.

From my laptop, the changes seems to be minor, what would you want to modify ?

Note: it seems that the logic I added to put the %%time label isn't working on some text cells (which explains why the previous example was failing)

PipKat commented on .gitignore:
```diff
@@ -381,6 +381,9 @@ test-output.xml
 # Scratch Jupyter Notebooks
 scratch_notebooks/
 
+# Ipython notebook checkpoints
```
```suggestion
# IPython notebook checkpoints
```

PipKat commented on doc/source/conf.py:
```diff
@@ -60,6 +67,18 @@ def run(self):
+# Sphinx event hooks
+
+def directory_size(directory_path):
+    """Compute the size (in mega bytes) of a directory."""
```
```suggestion
    """Compute the size in megabytes of a directory."""
```

PipKat commented on doc/source/conf.py:
```diff
@@ -42,6 +44,11 @@ def visit_desc_content(self, node: Element) -> None:
+EXAMPLES_DIRECTORY = path.resolve()
+
 class PrettyPrintDirective(Directive):
     """Renders a constant using ``pprint.pformat`` and inserts into the document."""
```
```suggestion
    """Renders a constant using ``pprint.pformat`` and inserts it into the document."""
```

PipKat commented on doc/source/conf.py:
```diff
@@ -60,6 +67,18 @@ def run(self):
+        for f in files:
+            fp = os.path.join(path, f)
+            res += os.stat(fp).st_size
+    # Convert in mega bytes
```
```suggestion
    # Convert in megabytes
```

PipKat commented on doc/source/conf.py:
```diff
@@ -70,19 +89,129 @@ def autodoc_skip_member(app, what, name, obj, skip, options):
+    logger.info(f"Doctree removed.")
+
+def copy_examples(app):
+    """Copy directory examples (root directory) files into the doc/source/examples directory.
```
```suggestion
    """Copy files in the ``examples`` root directory into the ``doc/source/examples`` directory.
```

PipKat commented on doc/source/conf.py:
```diff
@@ -70,19 +89,129 @@ def autodoc_skip_member(app, what, name, obj, skip, options):
+    logger.info(f"Copying examples from {EXAMPLES_DIRECTORY} to {DESTINATION_DIRECTORY}.")
+    if os.path.exists(DESTINATION_DIRECTORY):
+        size = directory_size(DESTINATION_DIRECTORY)
+        logger.info(f"Directory {DESTINATION_DIRECTORY} ({size} MB) already exist, removing it.")
```
```suggestion
        logger.info(f"Directory {DESTINATION_DIRECTORY} ({size} MB) already exists. Removing it.")
```

PipKat commented on doc/source/conf.py:
```diff
@@ -70,19 +89,129 @@ def autodoc_skip_member(app, what, name, obj, skip, options):
+    logger.info(f"Copy performed")
+
+def remove_examples(app, exception):
+    """Remove the doc/source/examples directory created during the documentation build.
```
```suggestion
    """Remove the ``doc/source/examples`` directory created during the documentation build.
```

PipKat commented on doc/source/conf.py:
```diff
@@ -70,19 +89,129 @@ def autodoc_skip_member(app, what, name, obj, skip, options):
+    # Get the full path to the document
+    docpath = os.path.join(app.srcdir, docname)
+
+    # Check if this is a .py example file
```
```suggestion
    # Check if this is a PY example file
```

PipKat commented on doc/source/conf.py:
```diff
@@ -70,19 +89,129 @@ def autodoc_skip_member(app, what, name, obj, skip, options):
+    if not os.path.exists(docpath + '.py') or not docname.startswith("examples"):
+        return
+
+    logger.info(f"Adding '# %%time' to file {docname}.py")
```
```suggestion
    logger.info(f"Adding '# %%time' to the `{docname}.py` file.")
```

PipKat commented on doc/source/conf.py:
```diff
@@ -70,19 +89,129 @@ def autodoc_skip_member(app, what, name, obj, skip, options):
+    # <img src="../../doc/source/_static/diff_via.png" width="500">
+    However, examples fail when used through the documentation build because
+    reaching the associated path should be "../../_static/diff_via.png".
+    Indeed, the directory ``_static`` is automatically copied into the output directory
```
```suggestion
    Indeed, the ``_static`` directory is automatically copied into the
```

PipKat commented on doc/source/conf.py:
```diff
@@ -70,19 +89,129 @@ def autodoc_skip_member(app, what, name, obj, skip, options):
+    source[0] = source[0].replace('../../doc/source/_static', '../../_static')
+
+def remove_ipython_time_from_html(app, pagename, templatename, context, doctree):
+    """Remove '# %%time' from examples generated HTML files.
```
```suggestion
    """Remove '# %%time' from examples generated HTML files.
```
Missing word? ... from examples generated [to/in/from?] HTML files."

PipKat commented on doc/source/conf.py:
```diff
@@ -70,19 +89,129 @@ def autodoc_skip_member(app, what, name, obj, skip, options):
+    """Remove '# %%time' from examples generated HTML files.
+    """
+    if pagename.startswith("examples") and not pagename.endswith("/index"):
+        logger.info(f"Removing '# %%time' from file {pagename}")
```
```suggestion
        logger.info(f"Removing '# %%time' from {pagename} file.")
```

PipKat commented on doc/source/conf.py:
```diff
@@ -236,6 +370,22 @@ def setup(app):
 # The name of the Pygments (syntax highlighting) style to use.
 pygments_style = "sphinx"
 
+# Execute notebooks before convertion
```
```suggestion
# Execute notebooks before conversion
```

PipKat commented on doc/source/conf.py:
```diff
@@ -236,6 +370,22 @@ def setup(app):
+# Execute notebooks before convertion
+nbsphinx_execute = "always"
+
+# Allow errors to help debug.
```
```suggestion
# Allow errors to help debug
```
Removing as other comments don't have concluding punctuation.

PipKat commented on doc/source/conf.py:
```diff
@@ -267,33 +417,33 @@ def setup(app):
     # necessary for pyvista when building the sphinx gallery
```
```suggestion
    # necessary for PyVista when building the Sphinx gallery
```

PipKat commented on doc/source/conf.py:
```diff
@@ -267,33 +417,33 @@ def setup(app):
+    #         "gallery_dirs": ["examples"],
+    #         # Pattern to search for examples files
+    #         "filename_pattern": r"\.py",
+    #         # Remove the "Download all examples" button from the top level gallery
```
```suggestion
    #         # Remove the "Download all examples" button from the top-level gallery
```

PipKat commented on doc/source/conf.py:
```diff
@@ -267,33 +417,33 @@ def setup(app):
+    #         "within_subsection_order": FileNameSortKey,
+    #         # directory where function granular galleries are stored
+    #         "backreferences_dir": None,
+    #         # Modules for which function level galleries are created.  In
```
```suggestion
    #         # Modules for which function-level galleries are created.  In
```

PipKat commented on examples/00-EDB/00_EDB_Create_VIA.py:
```diff
@@ -1,63 +1,46 @@
-# ~~~~~~~~~~~~~~~~~~
+# ## Add stackup layers
 # Add stackup layers.
 # A stackup can be created layer by layer or imported from a csv file or xml file.
```
```suggestion
# A stackup can be created layer by layer or imported from a CSV file or XML file.
```

PipKat commented on examples/00-EDB/00_EDB_Create_VIA.py:
```diff
@@ -71,22 +54,15 @@
-# Geometry Plot
-# ~~~~~~~~~~~~~
-#
+# ## Geometry Plot
```
```suggestion
# ## Geometry plot
```

PipKat commented on examples/00-EDB/00_EDB_Create_VIA.py:
```diff
@@ -71,22 +54,15 @@
-# ~~~~~~~~~~~~
-#
-edb.stackup.plot(plot_definitions="MyVia")
+# ## Stackup Plot
```
```suggestion
# ## Plot stackup
```
is the space after the first ``#`` a mistake.  The ``# ##`` seems odd.

PipKat commented on examples/00-EDB/psd_example.py:
```diff
@@ -0,0 +1,144 @@
+# Plotting power spectral density (PSD) using ``psd``.
+#
+# The PSD is a common plot in the field of signal processing. 
+# NumPy has many useful libraries for computing a PSD. Below we demo a few examples of 
+# how this can be accomplished and visualized with Matplotlib.
```
```suggestion
# NumPy has many useful libraries for computing a PSD. The following code provides a few examples of 
# how this can be accomplished and visualized with Matplotlib.
```
Avoid "we" and page positions (like below and above).

PipKat commented on examples/00-EDB/psd_example.py:
```diff
@@ -0,0 +1,144 @@
+
+import matplotlib.mlab as mlab
+
+# Fixing random state for reproducibility
```
```suggestion
# Fix random state for reproducibility
```

PipKat commented on examples/00-EDB/psd_example.py:
```diff
@@ -0,0 +1,144 @@
+# Equations:
+# $ E=mc^2 $
+#
+# Below we'll show a slightly more complex example that demonstrates how padding affects the resulting PSD.
```
```suggestion
# The follow code provides a slightly more complex example that demonstrates how padding affects the resulting PSD.
```

PipKat commented on examples/00-EDB/psd_example.py:
```diff
@@ -0,0 +1,144 @@
+axs['signal'].set_ylabel('Signal')
+
+# Plot the PSD with different amounts of zero padding. This uses the entire
+# time series at once
```
```suggestion
# time series at once.
```
Adding punctuation here because it is the second complete sentence.

PipKat commented on examples/00-EDB/psd_example.py:
```diff
@@ -0,0 +1,144 @@
+axs['zero padding'].psd(y, NFFT=len(t), pad_to=len(t) * 2, Fs=fs)
+axs['zero padding'].psd(y, NFFT=len(t), pad_to=len(t) * 4, Fs=fs)
+
+# Plot the PSD with different block sizes, Zero pad to the length of the
+# original data sequence.
```
```suggestion
# Plot the PSD with different block sizes and zero pad to the length of the
# original data sequence.
```

PipKat commented on examples/00-EDB/psd_example.py:
```diff
@@ -0,0 +1,144 @@
+# -
+
+# This is a ported version of a MATLAB example from the
+# signal processing toolbox that showed some difference at one time between
+# Matplotlib's and MATLAB's scaling of the PSD.
```
```suggestion
# signal processing toolbox that shows some differences at one time between
# Matplotlib's and MATLAB's scaling of the PSD.
```

PipKat commented on examples/00-EDB/pwm_spectrum.py:
```diff
@@ -0,0 +1,121 @@
+# # PWM Spectrum
+
+# This example illustrates the impact of switching rise and fall
+# time on the spectrum of a pulse-width modulated (PWM) signal.
```
```suggestion
# This example shows the impact of switching rise and fall
# time on the spectrum of a pulse-width modulated (PWM) signal.
```

PipKat commented on examples/00-EDB/pwm_spectrum.py:
```diff
@@ -0,0 +1,121 @@
+# Update the time series so each PWM cycle is fully sampled.  This simplifies the creation of the PMW
+# signal if the number of electrical cycles is not an integer multiple of the number of PWM cycles.
+#
+# The total number of time samples will be updated.
```
```suggestion
# The total number of time samples is updated.
```
Avoid future tense.

PipKat commented on examples/00-EDB/pwm_spectrum.py:
```diff
@@ -0,0 +1,121 @@
+t = np.linspace(0, t_tot, num=n_samples)
+print(f"Number of total samples: {n_samples}")
+print(f"Number of PWM cycles: {n_pwm_cycles}")
+print(f"Total time has been updated to {t_tot * 1E3:.1f} ms")
```
```suggestion
print(f"Total time has been updated to {t_tot * 1E3:.1f} ms.")
```

PipKat commented on examples/00-EDB/pwm_spectrum.py:
```diff
@@ -0,0 +1,121 @@
+
+n_pwm_cycles = int(np.max(t) * pwm_freq)
+print(f"Number of PWM cycles: {n_pwm_cycles}")
+print(f"Total time has been updated to {t_tot * 1E3:.1f} ms")
```
```suggestion
print(f"Total time has been updated to {t_tot * 1E3:.1f} ms.")
```

PipKat commented on examples/00-EDB/pwm_spectrum.py:
```diff
@@ -0,0 +1,121 @@
+
+f = lambda _freq, _t: 0.5 * (1 + np.sin(2 * np.pi * _freq * _t))  # Sinusoidal signal
+
+pwm = np.zeros(t.shape)  # Initialize the pwm data.
```
```suggestion
pwm = np.zeros(t.shape)  # Initialize the PWM data.
```

PipKat commented on examples/00-EDB/pwm_spectrum.py:
```diff
@@ -0,0 +1,121 @@
+lpf = lambda _w, _f: np.convolve(_w / _w.sum(), _f, mode="valid")
+spectrum = lambda _f: np.fft.fft(_f, norm="forward")
+
+# ## Rise and Fall time
```
```suggestion
# ## Rise and fall time
```

SMoraisAnsys commented:
Closing this PR as we re moving the examples into another repo.
